core/src/main/java/hudson/model/labels/LabelExpression.java

```diff
@@ -210,4 +225,158 @@ public LabelOperatorPrecedence precedence() {
             return LabelOperatorPrecedence.IMPLIES;
         }
     }
+
+    //region Auto-Completion and Validation
+
+    /**
+     * Utility class for taking the current input value and computing a list of potential terms to match against the
+     * list of defined labels.
+     */
+    static class AutoCompleteSeeder {
+        private String source;
+
+        AutoCompleteSeeder(String source) {
+            this.source = source;
+        }
+
+        List<String> getSeeds() {
+            ArrayList<String> terms = new ArrayList<>();
+            boolean trailingQuote = source.endsWith("\"");
+            boolean leadingQuote = source.startsWith("\"");
+            boolean trailingSpace = source.endsWith(" ");
+
+            if (trailingQuote || (trailingSpace && !leadingQuote)) {
+                terms.add("");
+            } else {
+                if (leadingQuote) {
+                    int quote = source.lastIndexOf('"');
+                    if (quote == 0) {
+                        terms.add(source.substring(1));
+                    } else {
+                        terms.add("");
+                    }
+                } else {
+                    int space = source.lastIndexOf(' ');
+                    if (space > -1) {
+                        terms.add(source.substring(space+1));
+                    } else {
+                        terms.add(source);
+                    }
+                }
+            }
+
+            return terms;
+        }
+    }
+
+    /**
+     * Plugins may want to contribute additional restrictions on the use of specific labels for specific jobs.
+     * This extension point allows such restrictions.
+     *
+     * @since TODO
+     */
+    public static abstract class LabelValidator implements ExtensionPoint {
```

Comment:
Could be an `interface` I suppose.

Reply:
Sure. I tried keeping it as close to the original as possible.

Reply:
Have done this now.